src/EditorFeatures/VisualBasic/NavigationBar/VisualBasicEditorNavigationBarItemService_CodeGeneration.vb

```diff
@@ -20,24 +20,29 @@ Imports Microsoft.VisualStudio.Text.Editor
 
 Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.NavigationBar
     Partial Friend Class VisualBasicEditorNavigationBarItemService
-        Private Sub GenerateCodeForItem(document As Document, generateCodeItem As AbstractGenerateCodeItem, textView As ITextView, cancellationToken As CancellationToken)
+        Private Async Function GenerateCodeForItemAsync(document As Document, generateCodeItem As AbstractGenerateCodeItem, textView As ITextView, cancellationToken As CancellationToken) As Task
             ' We'll compute everything up front before we go mutate state
-            Dim text = document.GetTextSynchronously(cancellationToken)
-            Dim newDocument = GetGeneratedDocumentAsync(document, generateCodeItem, cancellationToken).WaitAndGetResult(cancellationToken)
-            Dim generatedTree = newDocument.GetSyntaxRootSynchronously(cancellationToken)
+            Dim text = Await document.GetTextAsync(cancellationToken).ConfigureAwait(False)
+            Dim newDocument = Await GetGeneratedDocumentAsync(document, generateCodeItem, cancellationToken).ConfigureAwait(False)
+            Dim generatedTree = Await newDocument.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(False)
+
             Dim generatedNode = generatedTree.GetAnnotatedNodes(GeneratedSymbolAnnotation).Single().FirstAncestorOrSelf(Of MethodBlockBaseSyntax)
-            Dim documentOptions = document.GetOptionsAsync(cancellationToken).WaitAndGetResult(cancellationToken)
+            Dim documentOptions = Await document.GetOptionsAsync(cancellationToken).ConfigureAwait(False)
             Dim indentSize = documentOptions.GetOption(FormattingOptions.IndentationSize)
+
             Dim navigationPoint = NavigationPointHelpers.GetNavigationPoint(generatedTree.GetText(text.Encoding), indentSize, generatedNode)
 
+            ' switch back to ui thread to actually perform the application and navigation
+            Await Me.ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken)
+
             Using transaction = New CaretPreservingEditTransaction(VBEditorResources.Generate_Member, textView, _textUndoHistoryRegistry, _editorOperationsFactoryService)
                 newDocument.Project.Solution.Workspace.ApplyDocumentChanges(newDocument, cancellationToken)
 
-                NavigateToVirtualTreePoint(newDocument.Project.Solution, navigationPoint, cancellationToken)
+                Await NavigateToVirtualTreePointAsync(newDocument.Project.Solution, navigationPoint, cancellationToken).ConfigureAwait(True)
```

Comment:
It appears NavigateToVirtualTreePointAsync is only async because it'll SwitchToMainThread -- I'd encourage that function to stay sync and assert it's on the UI thread; here you just call the async method and that's fine. The one other place it's called just SwitchToMainThread in the caller.

Reply:
yup!  will do.